core-rust/core-api-server/src/core_api/browse.rs

```diff
@@ -745,30 +784,87 @@ impl<'s, S: SubstateDatabase> EngineStateDataLoader<'s, S> {
         let indexed_value = self
             .reader
             .read_object_field(node_id, module_id, field_meta.index.number)
-            .map_err(|error| match error {
-                SystemReaderError::NodeIdDoesNotExist => {
-                    EngineStateBrowsingError::RequestedItemNotFound(ItemKind::Entity)
-                }
-                SystemReaderError::ModuleDoesNotExist => {
-                    EngineStateBrowsingError::RequestedItemNotFound(ItemKind::Module)
-                }
-                SystemReaderError::FieldDoesNotExist => {
-                    EngineStateBrowsingError::RequestedItemNotFound(ItemKind::Field)
-                }
-                SystemReaderError::NotAnObject => EngineStateBrowsingError::RequestedItemInvalid(
-                    ItemKind::Entity,
-                    "not an object".to_string(),
-                ),
-                unexpected => EngineStateBrowsingError::UnexpectedEngineError(
-                    unexpected,
+            // if `ObjectFieldMeta` exists, then the object, module and field must exist - no errors expected:
+            .map_err(|error| {
+                EngineStateBrowsingError::UnexpectedEngineError(
+                    error,
                     "when reading object field".to_string(),
-                ),
+                )
             })?;
         Ok(SborData::new(
             indexed_value.into(),
             &field_meta.resolved_type,
         ))
     }
+
+    /// Returns an iterator over all keys of the given object's collection, starting from the given
+    /// key (or its successor, if it does not exist), in an arbitrary but deterministic order used
+    /// by the backing storage.
+    pub fn iter_object_collection_keys(
+        &self,
+        node_id: &NodeId,
+        module_id: ModuleId,
+        collection_meta: &'s ObjectCollectionMeta,
+        from_key: Option<&SubstateKey>,
+    ) -> Result<impl Iterator<Item = ObjectCollectionKey> + '_, EngineStateBrowsingError> {
+        let from_key = from_key.cloned();
+        Ok(self
+            .reader
+            .collection_iter(node_id, module_id, collection_meta.index.number)
+            // if `ObjectCollectionMeta` exists, then the object, module and collection must exist - no errors expected:
+            .map_err(|error| {
+                EngineStateBrowsingError::UnexpectedEngineError(
+                    error,
+                    "when reading object collection".to_string(),
+                )
+            })?
+            .map(|(substate_key, _)| substate_key)
+            // TODO(after adding "iterate from" support in Engine): The implementation below is a
```

Comment:
_(note from author)_

Blocked by https://github.com/radixdlt/radixdlt-scrypto/pull/1648
(but can easily be addressed in a separate PR)

Reply:
(the Scrypto dep has merged and I will address this TODO in the next PR)